url8012c8d031_dmbb966_BBMS_tgz-pJ8-testbed_JNEATtestJ8/src/jneat/Genome.java

```diff
@@ -0,0 +1,1528 @@
+		
+		// First, average traits
+		Vector<Trait> newTraits = AverageTraits(g);		
+		
+		// Second, determine which genome is better.
+		// The "worse" genome shouldn't be allowed to add extra structural baggage.
+		// If they are equally fit, then the smaller one's disjoint and excess genes will be used.
+
+		boolean p1better = false;
+		int size1 = genes.size();
+		int size2 = g.genes.size();
+		
+		if (fitness1 > fitness2) p1better = true;
+		else if (fitness1 == fitness2 && size1 < size2) p1better = true;
+				
+		Vector<Gene> newGenes = new Vector<Gene>();
+		Vector<NNode> newNodes = new Vector<NNode>();
+		
+		int j1 = 0;
+		int j2 = 0;
+		boolean skipGene = false;
+		boolean disableGene = false;
+		Gene chosenGene = null;
+		
+		while (j1 < size1 || j2 < size2) {
+			skipGene = false;		// Defaults to not skipping a chosen gene
+			disableGene = false;	
+			if (j1 >= size1) {
+				chosenGene = g.genes.elementAt(j2);
+				j2++;
+				if (p1better) skipGene = true;		// Skip excess from the worse genome
+			} else if (j2 >= size2) {
+				chosenGene = genes.elementAt(j1);
+				j1++;
+				if (!p1better) skipGene = true;		// Skip excess from the worse genome
+			} else {
+				Gene _p1Gene = genes.elementAt(j1);
+				Gene _p2Gene = g.genes.elementAt(j2);
+				
+				if (_p1Gene.innovation_num == _p2Gene.innovation_num) {
+					if (GlobalFuncs.randFloat() < 0.5) chosenGene = _p1Gene;
+					else chosenGene = _p2Gene;
+					
+					// If one of the genes is disabled, the corresponding gene in the offspring
+					// has a high chance of being disabled as well
+					if (!_p1Gene.enabled || !_p2Gene.enabled) {
+						if (GlobalFuncs.randFloat() < 0.75) disableGene = true;
+					}		
+					j1++;
+					j2++;	
+				} else if (_p1Gene.innovation_num < _p2Gene.innovation_num) {
+					chosenGene = _p1Gene;
+					j1++;
+					if (!p1better) skipGene = true;												
+				} else if (_p2Gene.innovation_num < _p1Gene.innovation_num) {
+					chosenGene = _p2Gene;
+					j2++;
+					if (p1better) skipGene = true;
+				}
+			}
+			
+			
+			if (!skipGene) skipGene = CheckGeneConflict(newGenes, chosenGene);			
+			
+			// Add gene if not skipped
+			if (!skipGene) {
+				AddGene(newNodes, newTraits, newGenes, chosenGene, disableGene);
+			} 
+		} // End while loop
+		
+		Genome newGenome = new Genome(newGenes, newTraits, newNodes, gID);
+		
+		boolean outputPresent = false;
+		
+		// Verify to ensure there are outputs
+		for (int i = 0; i < newNodes.size(); i++) {
+			NNode curNode = newNodes.elementAt(i);
+			if (curNode.gNodeLabel == NodeLabelEnum.OUTPUT) {
+				outputPresent = true;
+				break;
+			}
+		}
+		
+		if (!outputPresent) {
+			System.out.println("WARNING!  When conducting MateMultipoint, no output nodes found.");
+			System.out.println("Genome A:\n" + this.PrintGenome());
+			System.out.println("\nGenome B:\n" + g.PrintGenome());
+			System.out.println("\nResulting Genome:\n" + newGenome.PrintGenome());			
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 857:862, 950:955.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.